README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
genderizeR
==========

by Kamil Wais [homepage / contact](https://kalimu.github.io)

[![Licence](https://img.shields.io/badge/licence-MIT-blue.svg)](https://www.r-project.org/Licenses/MIT) [![Lifecycle](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://www.tidyverse.org/lifecycle/) [![Travis build status](https://travis-ci.org/kalimu/genderizeR.png?branch=master)](https://travis-ci.org/kalimu/genderizeR) [![CRAN Status](http://www.r-pkg.org/badges/version/genderizeR)](https://cran.r-project.org/package=genderizeR) [![CRAN Checks](https://cranchecks.info/badges/summary/genderizeR)](https://cran.r-project.org/web/checks/check_results_genderizeR.html) [![Monthly downloads badge](http://cranlogs.r-pkg.org/badges/last-month/genderizeR)](https://cran.r-project.org/package=genderizeR) [![Daily downloads badge](https://cranlogs.r-pkg.org/badges/last-day/genderizeR?color=blue)](https://CRAN.R-project.org/package=genderizeR) [![Weekly downloads badge](https://cranlogs.r-pkg.org/badges/last-week/genderizeR?color=blue)](https://CRAN.R-project.org/package=genderizeR) [![HitCount](http://hits.dwyl.io/kalimu/genderizer.svg)](http://hits.dwyl.io/kalimu/genderizer)

R package for gender predictions based on first names.

The package home page: <https://kalimu.github.io/project/genderizer/>

Information about the genderize.io project and documentation of the API: <http://genderize.io>

Description
-----------

The genderizeR package uses genderize.io API to **predict gender** from first names extracted from text corpus (not only from clean vectors of given names). The accuracy of **prediction could be controlled** by two parameters: counts of first names in database and probability of gender given the first name. The package has also built-in functions that can **calculate specific errors** (also with bootstrapping), **train algorithm** on training dataset (with gender labels) and **prepare character vectors for gender checking**.

Installing the package
----------------------

### Installing stable version from CRAN

``` r
install.packages('genderizeR')
```

### Installing developer version from GitHub

Remember to install `devtools` package first!

``` r
# install.packages('devtools')
devtools::install_github("kalimu/genderizeR")
```

### Loading the installed package

``` r
library(genderizeR)
```

    #> 
    #> Welcome to genderizeR package version: 2.0.0.9003
    #> 
    #> Homepage: http://www.wais.kamil.rzeszow.pl/genderizeR
    #> 
    #> Changelog: news(package = 'genderizeR')
    #> Help & Contact: help(genderizeR)
    #> 
    #> If you find this package useful cite it please. Thank you!
    #> See: citation('genderizeR')
    #> 
    #> To suppress this message use:
    #> suppressPackageStartupMessages(library(genderizeR))

A working example
-----------------

``` r
# An example for a character vector of strings
x = c("Winston J. Durant, ASHP past president, dies at 84",
"JAN BASZKIEWICZ (3 JANUARY 1930 - 27 JANUARY 2011) IN MEMORIAM",
"Maria Sklodowska-Curie")
 
# Search for terms that could be first names
# If you have your API key you can authorize access to the API with apikey argument
# e.g. findGivenNames(x, progress = FALSE, apikey = 'your_api_key')
givenNames = findGivenNames(x, progress = FALSE)
```

``` r

# Use only terms that have more than x counts in the database
givenNames = givenNames[count > 100]
givenNames
#>       name gender probability count
#> 1: winston   male        0.98   128
#> 2:     jan   male         0.6  1663
#> 3:   maria female        0.99  8402

# Genderize the original character vector
genderize(x, genderDB = givenNames, progress = FALSE)
#>                                                              text
#> 1:             Winston J. Durant, ASHP past president, dies at 84
#> 2: JAN BASZKIEWICZ (3 JANUARY 1930 - 27 JANUARY 2011) IN MEMORIAM
#> 3:                                         Maria Sklodowska-Curie
#>    givenName gender genderIndicators
#> 1:   winston   male                1
#> 2:       jan   male                1
#> 3:     maria female                1
```

Tutorial
--------

For more comprehensive tutorial check the vignette in the package.

``` r
browseVignettes("genderizeR")
```

What's new in the package?
--------------------------

``` r
news(package = 'genderizeR')
```

See package help pages in R / Rstudio
-------------------------------------

``` r
help(package = 'genderizeR')
?textPrepare
?findGivenNames
?genderize
```

How to contribute to the package?
---------------------------------

### For bugs, updates and new functionalities:

Fork git repo `https://github.com/kalimu/genderizeR` and submit a pull request.

### Feedback:

If you enjoy using the package you could write a short testimonial and send it to me. I will be happy to post in on the package homepage.

For any kind of feedback you can use the contact form here: <https://kalimu.github.io/#contact>

How to contact the package's author regarding research or commercial project?
-----------------------------------------------------------------------------

Please use the contact form: <https://kalimu.github.io/#contact>

How to cite the package?
------------------------

``` r
citation('genderizeR')
#> 
#> Wais K (2006). "Gender Prediction Methods Based on First Names
#> with genderizeR." _The R Journal_, *8*(1), 17-37. doi:
#> 10.32614/RJ-2016-002 (URL: http://doi.org/10.32614/RJ-2016-002),
#> <URL: https://doi.org/10.32614/RJ-2016-002>.
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     title = {{Gender Prediction Methods Based on First Names with
#>           genderizeR}},
#>     author = {Kamil Wais},
#>     year = {2006},
#>     journal = {{The R Journal}},
#>     doi = {10.32614/RJ-2016-002},
#>     pages = {17--37},
#>     volume = {8},
#>     number = {1},
#>     url = {https://doi.org/10.32614/RJ-2016-002},
#>   }
```

Thank You for the citation!


NEWS
# genderizeR 2.1.1

## Major changes  
* function findGivenNames() returns now a data frame always with additional column "country_id"

## Minor changes
* more robust unit tests

# genderizeR 2.0.1

## Major changes  
* new parameters for findGivenNames() and genderizeAPI() ("country" and "language") 
which allow for localized searches (thanks to @tklebel)

## Minor changes
* corrected typos and formulation in documentation files
* depracted function names are removed
* examples from help files are all chcecked to be working with the current package version
* updated DESCRIPTION file
* new CITATION file see citation('genderizeR')

# genderizeR 2.0.0

* parallel processing in genderizeTrain works properly under Linux now
* simple solution for caching results for findGivenNames() 
* fix in misspelled name function (classificationErrors)
* small improvement in textPrepare() function
* updated and corrected help pages and function examples
* more comprehensive exemplary datasets description
* new more comprehensive set of unit tests
* small fixes and code cleaning 
* more thorough unit tests
* updated README file
* added tutorial vignette

# genderizeR 1.2.1

* fix in function textPrepare (now works properly with initials at the end)
* documentation corrected
* code cleared

# genderizeR 1.2.0

* fix for changes in version 1.0.0 of httr package

# genderizeR 1.1.1

* small DESCRIPTION update

# genderizeR 1.1.0

* second CRAN release working with new API now
* RCurl package replaced by httr
* https connection fixed
* default queryLength set to 10
* small bug fixes
* implemented API authorization
* simple unit tests
* workaround with multi handle error when the API query is stopped by user
* added ssl.verifypeer option

# genderizeR 1.0.0.3

* ncol error fixed
* 'like' term works

# genderizeR 1.0.0.1

* https temporary fix

# genderizeR 1.0.0.0

* hotfixes and updates
* first CRAN release

# genderizeR 0.1.2.4

* training function with parallel version as well


# genderizeR 0.1.2.3

* dataset with titles sample
* dataset with first names and their gender data for the titles sample


# genderizeR 0.1.2.2

* dataset with authorships sample 
* dataset with first names and their gender data for the authorships sample

# genderizeR 0.1.2.1

* improved progress monitoring 


# genderizeR 0.1.1

* optional distributed corpus in textPrepare function


# genderizeR 0.0.1

* dev version building on GitHub


R/genderizeAPI.R
#' Getting data from genderize.io  API
#' 
#' The \code{genderizeAPI} function connects to genderize.io API and checks if 
#' a term (one or more) is in the genderize.io database and returns 
#' predicted gender probability and count of the records with this 
#' term in the database.
#' 
#' 
#' @param x A vector of terms to check in genderize.io database.
#' @param country A character string with a country code for localized search
#' of names. Country codes follow the ISO_3166-1 alpha-2 standard
#' \url{https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2}.
#' @param language A character string with a language code for localized search
#' of names. Language codes follow the ISO_639-1 standard: 
#' \url{https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes}
#' @param apikey A character string with the API key obtained via 
#' \url{https://store.genderize.io}. A default is NULL, which uses the free API 
#'  plan. If you reached the limit of the API you can start from the last 
#'  checked term next time.
#' @param ssl.verifypeer If TRUE (default) it checks the SSL Certificate. 
#'
#' @return A list of four elements: \code{response} is a data frame with names, 
#' genders, probabilities and counts or \code{NULL} if no terms are found 
#' in the genderize.io database; \code{limitLeft} is showing how many API queries 
#' are still possible within the current \code{limit} which will be renewed 
#' in \code{limitReset} seconds.
#' 
#' 
#' 
#' @examples 
#' \dontrun{
#' 
#' terms = c("loremipsum")
#' genderizeAPI(terms)$response
#' # Null data.table (0 rows and 0 cols)
#' 
#' terms = c("jan", "maria", "norbert", "winston", "loremipsum")
#' genderizeAPI(terms)
#'  
#' # example of the function output 
#' $response
#'       name gender probability count
#'       1:     jan   male        0.60  1692
#'       2:   maria female        0.99  8467
#'       3: norbert   male        1.00    77
#'       4: winston   male        0.98   128

#' 
#' $limitLeft
#' [1] 967
#' 
#' $limit
#' [1] 1000
#' 
#' $limitReset
#' [1] 83234 
#' 
#' }
#' 
#' @export

 
genderizeAPI = function(x,
                        country = NULL,
                        language = NULL,
                        apikey = NULL, 
                        ssl.verifypeer = TRUE
                        ) {
    
    termsQuery = x
    
    query = as.list(termsQuery)
    
    names(query)  = paste0('name[', 0:(length(termsQuery) - 1), ']')
    
    if (!is.null(country)) {
      
      query = c(query, 'country_id' = country)

    }
    
    if (!is.null(language)) {

      query = c(query, 'language_id' = language)
    
    }

    
    if (!is.null(apikey)) {
        
        query = c('apikey' = apikey, query)
        
    }
     
    # fix for version 1.0.0 of the httr package        
    #  r = httr::GET("https://api.genderize.io", httr::timeout(100), 
    #  query = query, httr::config(ssl.verifypeer = ssl.verifypeer)) 
    
    r = httr::GET("https://api.genderize.io", query = query, 
                  httr::config(ssl_verifypeer = ssl.verifypeer))
    
    if (httr::status_code(r) == 200) {
        
        l = httr::content(r)
        if (is.atomic(l[[1]])) {l = list(l)}
        l = l[unlist(lapply(l, function(x) {!is.null(x$gender)}))]
        
        limitLeft = as.numeric(httr::headers(r)$'x-rate-limit-remaining')
        limit = as.numeric(httr::headers(r)$'x-rate-limit-limit')
        limitReset = as.numeric(httr::headers(r)$'x-rate-reset')
        
        return(
            list(
                response = data.table::rbindlist(l),
                limitLeft = limitLeft, 
                limit = limit, 
                limitReset = limitReset
                )
            )
        
    } else {
        
        cat('\n', httr::http_status(r)$message)
        cat('\n', httr::content(r)$error)
        
        if (httr::status_code(r) == 429) {
            
            warning('You have used all available requests in this subscription plan.')
            
        }
        
        
        return
    }
    
}

